File: Client/Linux/lib/Management/UserManagementClient.cpp
#include <stdio.h>
#include <iostream>
#include <string.h>
#include "ManagementClient.h"
#include "../RestConnection.h"
#include "ArduinoJson.h"
#include "common.h"

using namespace std;

namespace Thriot { namespace Management {

/**
Initializes a new instance of the user management client

@param restConnection Wrapper class for the rest API calls
*/
UserManagementClient::UserManagementClient(RestConnection* restConnection) 
{
	_restConnection = restConnection;
	_isLoggedIn = false;
}

/**
Register a new user. 
Return 0 if there was no error, -1 if email activation is needed, HTTP status code on other error.
On successfull registration all subsequent operations will be executed in the name of the registered user.

@param reg Registration parameters

@return operation status. 0 if Ok.*/
int UserManagementClient::Register(const RegisterInfo& reg)
{
	DynamicJsonBuffer jsonBufferRequest;
	JsonObject& regJson = jsonBufferRequest.createObject();

	regJson["Name"] = reg.Name.c_str();
	regJson["Email"] = reg.Email.c_str();
	regJson["Password"] = reg.Password.c_str();

	char jsonBuffer[256];
	regJson.printTo(jsonBuffer, sizeof(jsonBuffer));

	Response httpResponse =_restConnection->Post("users/register", "application/json", jsonBuffer);

	if(httpResponse.Code != 200)
		return httpResponse.Code;

	DynamicJsonBuffer jsonBufferResponse;
	JsonObject& respJson = jsonBufferResponse.parseObject((char *)httpResponse.Body.c_str());
	
	bool needsActivation = (bool)respJson["NeedsActivation"];
	
	if(needsActivation)
		return -1;

	_isLoggedIn = true;

	return 0;
}

/**
Login a new user.
On successfull login all subsequent operations will be executed in the name of the logged in user.
If everything was fine 0 is returned otherwise a http status code.

@param login Login user email and password

@return operation status. 0 if Ok*/
int UserManagementClient::Login(const LoginInfo& login)
{
	DynamicJsonBuffer jsonBufferRequest;

	JsonObject& regJson = jsonBufferRequest.createObject();

	regJson["Email"] = login.Email.c_str();
	regJson["Password"] = login.Password.c_str();

	char jsonBuffer[256];
	regJson.printTo(jsonBuffer, sizeof(jsonBuffer));

	Response httpResponse =_restConnection->Post("users/login", "application/json", string(jsonBuffer));

	if(httpResponse.Code != 204)
		return httpResponse.Code;

	_isLoggedIn = true;

	return 0;
}

/**
Activate the user.
Generaly this method wouldn't be called from the client sinde the activation link is embedded into the
email intended to force user activation.

@param activate Activation parameters (userid and activation code)

@return operation status. 0 if Ok*/
int UserManagementClient::Activate(const ActivateInfo& activate)
{
	Response httpResponse =_restConnection->Get("users/activate/" + activate.UserId + "/" + activate.ActivationCode);

	if(httpResponse.Code != 204)
		return httpResponse.Code;

	_isLoggedIn = true;

	return 0;
}

/**
Resend activation email
When the activation email didn't arrive this method can be used to resend it.

@param email Email address to resend activation email to

@return operation status 0 if Ok */
int UserManagementClient::ResendActivationEmail(const string& email)
{
	DynamicJsonBuffer jsonBufferRequest;

	JsonObject& regJson = jsonBufferRequest.createObject();

	regJson["Email"] = email.c_str();

	char jsonBuffer[256];
	regJson.printTo(jsonBuffer, sizeof(jsonBuffer));

	Response httpResponse =_restConnection->Post("users/resendActivationEmail", "application/json", string(jsonBuffer));

	if(httpResponse.Code != 204)
		return httpResponse.Code;

	return 0;
}

/**
Send reset forgot password email
When the user forgot the password send an email with a link that can be used to reset the password.

@param email Email address to sent forgot password email

@return operation status 0 if Ok */
int UserManagementClient::SendForgotPasswordEmail(const string& email)
{
	DynamicJsonBuffer jsonBufferRequest;

	JsonObject& regJson = jsonBufferRequest.createObject();

	regJson["Email"] = email.c_str();

	char jsonBuffer[256];
	regJson.printTo(jsonBuffer, sizeof(jsonBuffer));

	Response httpResponse =_restConnection->Post("users/sendForgotPasswordEmail", "application/json", string(jsonBuffer));

	if(httpResponse.Code != 204)
		return httpResponse.Code;

	return 0;
}

/**
Reset password. 
This method generaly employs has the parameters that the SendForgotPasswordEmail contained

@param resetPassword Class containing reset password info

@return operation status 0 if Ok */
int UserManagementClient::ResetPassword(const ResetPasswordInfo& resetPassword)
{
	DynamicJsonBuffer jsonBufferRequest;

	JsonObject& regJson = jsonBufferRequest.createObject();

	regJson["UserId"] = resetPassword.UserId.c_str();
	regJson["ConfirmationCode"] = resetPassword.ConfirmationCode.c_str();
	regJson["Password"] = resetPassword.Password.c_str();

	char jsonBuffer[256];
	regJson.printTo(jsonBuffer, sizeof(jsonBuffer));

	Response httpResponse =_restConnection->Post("users/resetPassword", "application/json", string(jsonBuffer));

	if(httpResponse.Code != 204)
		return httpResponse.Code;

	return 0;
}

/**
Change password. 
Change password of the already logged in user to use a different password in the future.

@param changePassword Class containing change password info

@return operation status 0 if Ok */
int UserManagementClient::ChangePassword(const ChangePasswordInfo& changePassword)
{
	DynamicJsonBuffer jsonBufferRequest;

	JsonObject& regJson = jsonBufferRequest.createObject();

	regJson["CurrentPassword"] = changePassword.CurrentPassword.c_str();
	regJson["NewPassword"] = changePassword.NewPassword.c_str();

	char jsonBuffer[256];
	regJson.printTo(jsonBuffer, sizeof(jsonBuffer));

	Response httpResponse =_restConnection->Post("users/changePassword", "application/json", string(jsonBuffer));

	if(httpResponse.Code != 204)
		return httpResponse.Code;

	return 0;
}

/** Logoff the currently logged-in user */
void UserManagementClient::Logoff()
{
	_restConnection->Post("users/logoff", "application/json", "");

	_isLoggedIn = false;
}

/**
Determines whether the user is logged in

@return user login status*/
bool UserManagementClient::IsLoggedIn()
{
	return _isLoggedIn;
}

/**
Get the currently logged in user.
If the operation is successfull a valid user object is returned otherwise an uninitialized one.

@return User object*/
User UserManagementClient::Get()
{
	Response httpResponse =_restConnection->Get("users/me");

	User user;
	
	if(httpResponse.Code != 200)
		return user;

	DynamicJsonBuffer jsonBufferResponse;
	JsonObject& respJson = jsonBufferResponse.parseObject((char *)httpResponse.Body.c_str());
	
	user.Id = string(respJson["Id"].asString());
	user.Name = string(respJson["Name"].asString());
	user.Email = string(respJson["Email"].asString());

	return user;
}

/**
Find user by email address.
If the operation is successfull a valid user object is returned otherwise an uninitialized one.

@param email email address of the user we are looking for 

@return User object*/
User UserManagementClient::FindUser(const string& email)
{
	Response httpResponse =_restConnection->Get("users/byemail/" + UrlEncode(email) + "/");

	User user;
	
	if(httpResponse.Code != 200)
		return user;

	DynamicJsonBuffer jsonBufferResponse;
	JsonObject& respJson = jsonBufferResponse.parseObject((char *)httpResponse.Body.c_str());
	
	user.Id = string(respJson["Id"].asString());
	user.Name = string(respJson["Name"].asString());
	user.Email = string(respJson["Email"].asString());

	return user;
}
}}

File: Client/DotNet/Thriot.Client.DotNet/Management/UserManagementClient.cs
using System.Collections.Generic;
using System.Net;
using System.Web;

namespace Thriot.Client.DotNet.Management
{
    /// <summary>
    /// This class is responsible for managing users and doing login operations. Activation is not currently the part of this library.
    /// </summary>
    public class UserManagementClient : SpecificManagementClient
    {
        private readonly string _baseUrl;
        private bool _isLoggedIn;

        internal UserManagementClient(string baseUrl, IRestConnection restConnection) : base(restConnection)
        {
            _baseUrl = baseUrl;
            _isLoggedIn = false;
        }

        /// <summary>
        /// Register a new user in the system. At minimum the email address, the full name and the password is required at registration.
        /// If the system is configured not to require user activation after registration the function also logs in the user just registered.
        /// However if activation is required the <see cref="ActivationRequiredException" /> exeption will be thrown with the following message
        /// "Activation needed, please confirm your account by clicking the link you received by email" but the user will be registered. 
        /// It meens that the user registration happened but the system cannot be used unless clicking the activation link.
        /// 
        /// Send POST request to APIROOT/users/register
        /// </summary>
        /// <param name="register">Registration parameters</param>
        /// <exception cref="ActivationRequiredException">This exception indicates that the registration succeeded but activation is required before using the system</exception>
        /// <exception cref="WebException">This exception will be thrown in case of service side registration error. Please refer to the HTTP error code for more information</exception>
        public void Register(Register register)
        {
            var response = RestConnection.Post("users/register", JsonSerializer.Serialize(register));

            var registrationResult = JsonSerializer.Deserialize<RegistrationResult>(response);

            if (registrationResult.NeedsActivation)
                throw new ActivationRequiredException(
                    "Activation needed, please confirm your account by clicking the link you received by email");

            RestConnection.Setup(_baseUrl,
                new Dictionary<string, string>
                {
                    {
                        HttpRequestHeader.Authorization.ToString(), "Basic " + registrationResult.AuthToken
                    }
                });
            _isLoggedIn = true;
        }

        /// <summary>
        /// Log in the current user. On successfull login all subsequent management operations will run in the context of the logged in user.
        /// 
        /// Send POST request to APIROOT/users/login
        /// </summary>
        /// <param name="login">The user's email address and password</param>
        /// <exception cref="WebException">In case of any service side error an exception will be thrown. Please refer to the HTTP error code for more information</exception>
        public void Login(Login login)
        {
            var response = RestConnection.Post("users/login", JsonSerializer.Serialize(login));

            RestConnection.Setup(_baseUrl,
                new Dictionary<string, string>
                {
                    {
                        HttpRequestHeader.Authorization.ToString(), "Basic " + JsonSerializer.Deserialize<string>(response)
                    }
                });
            _isLoggedIn = true;
        }

        /// <summary>
        /// Activate the given user with the activation code
        /// 
        /// Send GET request to APIROOT/users/activate/{userId}/{activationCode}
        /// </summary>
        /// <param name="activate">Identifier of the user secured by the activation code to activate</param>
        public void Activate(Activate activate)
        {
            RestConnection.Get(string.Format("users/activate/{0}/{1}", activate.UserId, activate.ActivationCode));
        }

        /// <summary>
        /// Resend activation email for the given email address
        /// 
        /// Send POST request to APIROOT/users/resendActivationEmail
        /// </summary>
        /// <param name="email">Email address to send activation email for</param>
        public void ResendActivationEmail(EmailWrapper email)
        {
            RestConnection.Post("users/resendActivationEmail",
                JsonSerializer.Serialize(email));
        }

        /// <summary>
        /// Send email about forgotten password to the user having the email address in the parameter
        /// 
        /// Send POST request to APIROOT/users/sendForgotPasswordEmail
        /// </summary>
        /// <param name="email">Email address to send forgot password email for</param>
        public void SendForgotPasswordEmail(EmailWrapper email)
        {
            RestConnection.Post("users/sendForgotPasswordEmail",
                JsonSerializer.Serialize(email));
        }

        /// <summary>
        /// Reset the password specifid by the resetPassword dto. It will contain the userid, the confirmation code and the new password
        /// The first two parameters are sent out by the sendForgotPasswordEmail operation
        /// 
        /// Send POST request to APIROOT/users/resetPassword
        /// </summary>
        /// <param name="resetPassword">Parameters for resetting password</param>
        public void ResetPassword(ResetPassword resetPassword)
        {
            RestConnection.Post("users/resetPassword",
                JsonSerializer.Serialize(resetPassword));
        }

        /// <summary>
        /// Change password of the currently logged in
        /// You must specify the current and the new passwords
        /// 
        /// Send POST request to APIROOT/users/changePassword
        /// </summary>
        /// <param name="changePassword">Change password parameters</param>
        public void ChangePassword(ChangePassword changePassword)
        {
            RestConnection.Post("users/changePassword",
                JsonSerializer.Serialize(changePassword));
        }

        /// <summary>
        /// Query the currently logged in user
        /// 
        /// Send GET request to APIROOT/users/me
        /// </summary>
        /// <returns>Current user properties like full name and email address</returns>
        /// <exception cref="WebException">In case of any service side error an exception will be thrown. Please refer to the HTTP error code for more information</exception>
        public User Get()
        {
            var userStr = RestConnection.Get("users/me");

            return JsonSerializer.Deserialize<User>(userStr);
        }

        /// <summary>
        /// Find an user by email address. Only exact matches will work while the comparison is implemented case insensitively on the service side.
        /// 
        /// Send GET request to APIROOT/users/byemail/UrlEncode(email)/
        /// </summary>
        /// <param name="email">Email address to look for</param>
        /// <returns>User entity if the email address matches otherwise a null will be returned.</returns>
        /// <exception cref="WebException">In case of any service side error an exception will be thrown. Please refer to the HTTP error code for more information</exception>
        public User FindUser(string email)
        {
            var userStr = RestConnection.Get("users/byemail/" + HttpUtility.UrlEncode(email) + "/");

            return JsonSerializer.Deserialize<User>(userStr);
        }

        /// <summary>
        /// Logs off the user. Any subsequent management operations will fail that require authentication.
        /// </summary>
        public void Logoff()
        {
            RestConnection.Setup(_baseUrl, null);
            _isLoggedIn = false;
        }

        /// <summary>
        /// Indicates if anybody is logged in.
        /// </summary>
        public bool IsLoggedIn
        {
            get { return _isLoggedIn; }
        }
    }
}
